drivers/bme280/bme280.c

```diff
@@ -0,0 +1,453 @@
+ */
+static uint8_t measurement_regs[8];
+
+/*---------------------------------------------------------------------------*
+ *                          BME280 Core API                                  *
+ *---------------------------------------------------------------------------*/
+
+int bme280_init(bme280_t* dev, i2c_t i2c, uint8_t addr, const bme280_setting_t* settings)
+{
+    uint8_t chip_id;
+
+    dev->i2c_dev = i2c;
+    dev->i2c_addr = addr;
+    dev->settings = *settings;
+
+    /* Initialize I2C interface */
+    if (i2c_init_master(dev->i2c_dev, I2C_SPEED_NORMAL)) {
+        DEBUG("[Error] I2C device not enabled\n");
+        return -1;
+    }
+
+    /* Read chip ID */
+    chip_id = read_u8_reg(dev, BME280_CHIP_ID_REG);
+    if (chip_id != 0x60) {
+        DEBUG("[Error] Did not detect a BME280 at address %02x (%02X != %02X)\n",
+              dev->i2c_addr, chip_id, 0x60);
+        return -2;
+    }
+
+    /* Read compensation data, 0x88..0x9F, 0xA1, 0xE1..0xE7 */
+    if (read_calibration_data(dev)) {
+        DEBUG("[Error] Could not read calibration data\n");
+        return -3;
+    }
+
+    return 0;
+}
+
+void bme280_auto_init(void)
+{
+    for (unsigned i = 0; i < BME280_NUMOF; i++) {
+        if (bme280_init(&bme280_devs[i], bme280_params[i].i2c_dev,
+                        bme280_params[i].i2c_addr,
+                        &bme280_params[i].settings) < 0) {
+            LOG_ERROR("Unable to initialize BME280 sensor #%i\n", i);
+        }
+    }
+}
+
+/*
+ * Returns temperature in DegC, resolution is 0.01 DegC.
+ * t_fine carries fine temperature as global value
+ */
+int bme280_read_temperature(bme280_t* dev, float* temperature)
+{
+    if (!dev) {
+        return -1;
+    }
+    if (temperature) {
+        *temperature = 0;
+    }
+
+    if (do_measurement(dev) < 0) {
+        return -1;
+    }
+
+    bme280_calibration_t *cal = &dev->calibration;      /* helper variable */
+
+    /* Read the uncompensated temperature */
+    int32_t adc_T = (((uint32_t)measurement_regs[3 + 0]) << 12) |
+        (((uint32_t)measurement_regs[3 + 1]) << 4) |
+        ((((uint32_t)measurement_regs[3 + 2]) >> 4) & 0x0F);
+
+    /*
+     * Compensate the temperature value.
+     * The following is code from Bosch's BME280_driver bme280_compensate_temperature_int32()
+     * The variable names and the many defines have been modified to make the code
+     * more readable.
+     */
+    int32_t var1;
+    int32_t var2;
+
+    var1 = ((((adc_T >> 3) - ((int32_t)cal->dig_T1 << 1))) * ((int32_t)cal->dig_T2)) >> 11;
+    var2 = (((((adc_T >> 4) - ((int32_t)cal->dig_T1)) * ((adc_T >> 4) - ((int32_t)cal->dig_T1))) >> 12) *
+            ((int32_t)cal->dig_T3)) >> 14;
+
+    /* calculate t_fine (used for pressure and humidity too) */
+    t_fine = var1 + var2;
+
+    float output = (t_fine * 5 + 128) >> 8;
+
+    /* Convert to degrees Celcius */
+    output = output / 100;
```

Comment:
That's fine with me.

Notice that there are three sensor values: temperature, humidity and pressure. I'll make sure that all three of them don't use floating point arithmetic.
The code uses 64 bits integer arithmetic, including multiplication, and one 64 bits division. That will cause an increment of code size. But I don't know how to avoid that. BTW. The code is basically from a reference library supplied by Bosch on Github.